feat: add C ndarray API and refactor `blas/ext/base/dsorthp` (#5018)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/ext/base/dsorthp/README.md

-   **N**: number of indexed elements.
-   **order**: sort order. If `order < 0.0`, the input strided array is sorted in **decreasing** order. If `order > 0.0`, the input strided array is sorted in **increasing** order. If `order == 0.0`, the input strided array is left unchanged.
-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length.

> The `N` and stride parameters determine which elements in `x` are accessed at runtime. For example, to sort every other element

Reviewer: ```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to sort every other element:
```

The function has the following additional parameters:

-   **offsetX**: starting index.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to access only the last three elements:

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on starting a index. For example, to access only the last three elements:
```

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/dsorthp.h"
```

#### stdlib_strided_dsorthp( N, order, \*X, strideX )

Sorts a double-precision floating-point strided array using heapsort.

```c
double x[] = { 1.0, -2.0, 3.0, -4.0 };

> stdlib_strided_dsorthp( 2, -1, x, 1 );

Reviewer: ```suggestion
stdlib_strided_dsorthp( 2, -1.0, x, 1 );
```

```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **order**: `[in] double` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
-   **X**: `[inout] double*` input array.

> -   **strideX**: `[in] CBLAS_INT` stride length for `X`.

Reviewer: ```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length.
```

```c

> stdlib_strided_dsorthp( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX );

Reviewer: ```suggestion
stdlib_strided_dsorthp( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX );
```

```

<!--lint disable maximum-heading-length-->

#### stdlib_strided_dsorthp_ndarray( N, order, \*X, strideX, offsetX )

<!--lint enable maximum-heading-length-->

Sorts a double-precision floating-point strided array using heapsort and alternative indexing semantics.

```c
double x[] = { 1.0, -2.0, 3.0, -4.0 };

> stdlib_strided_dsorthp_ndarray( 4, 1, x, 1, 0 );

Reviewer: ```suggestion
stdlib_strided_dsorthp_ndarray( 4, 1.0, x, 1, 0 );
```

```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.

> -   **order**: `[in] CBLAS_INT` sort order.

Reviewer: ```suggestion
-   **order**: `[in] double` sort order.
```

> -   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.

Reviewer: ```suggestion
-   **X**: `[inout] double*` input array.
```

> -   **strideX**: `[in] CBLAS_INT` stride length for `X`.

Reviewer: ```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length.
```

> -   **offsetX**: `[in] CBLAS_INT` starting index for `X`.

Reviewer: ```suggestion
-   **offsetX**: `[in] CBLAS_INT` starting index.
```

```c

> stdlib_strided_dsorthp_ndarray( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX );

Reviewer: ```suggestion
stdlib_strided_dsorthp_ndarray( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```
